extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/IncrementalPublishingKafkaIndexTaskRunner.java

```diff
@@ -576,15 +561,19 @@ public void onFailure(Throwable t)
 
               nextOffsets.put(record.partition(), record.offset() + 1);
             }
-
-            if (nextOffsets.get(record.partition()).equals(endOffsets.get(record.partition()))
+            if (nextOffsets.get(record.partition()) >= endOffsets.get(record.partition())
                 && assignment.remove(record.partition())) {
               log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
               KafkaIndexTask.assignPartitions(consumer, topic, assignment);
               stillReading = !assignment.isEmpty();
             }
           }
-
+          if (nextOffsets.get(currentPartition) != null
```

Comment:
I think this should be placed inside the loop, otherwise it only updates the `nextOffsets` of at most 1 partition. In the case where the polled `records` contain records of more than 1 partition, the nextOffsets of all those partitions should be updated.

Reply:
Thanks, inside the loop `nextOffsets` is updated by `nextOffsets.put(record.partition(), record.offset() + 1);` That said, let me double check, I'm a little rusty on this PR now.

Reply:
Sorry for the late reply. If I understand correctly, the line `nextOffsets.put(currentPartition, endOffsets.get(currentPartition));` is supposed to update the `nextOffsets` of the partitions with polled offsets that have exceeded the partition's `endOffsets`. But what if, in the list of polled `records`, 2 or more partitions have offsets that exceed their respective endOffsets. Then since `nextOffsets.put(currentPartition, endOffsets.get(currentPartition));` is outside the loop, it would only update the `nextOffsets` of 1 such partition. I remember talking to @jihoonson about this, he may know more.